src/main/io/rcdevice.h

```diff
@@ -169,40 +111,63 @@ typedef struct runcamDevice_s {
     serialPort_t *serialPort;
     uint8_t buffer[RCDEVICE_PROTOCOL_MAX_PACKET_SIZE];
     runcamDeviceInfo_t info;
+    bool isReady;
 } runcamDevice_t;
 
-typedef struct runcamDeviceSettingTextSelection_s {
-    char text[RCDEVICE_PROTOCOL_MAX_SETTING_VALUE_LENGTH];
-} runcamDeviceSettingTextSelection_t;
-
-typedef struct runcamDeviceSettingDetail_s {
-    uint8_t type;
-    uint32_t value;
-    uint32_t minValue;
-    uint32_t maxValue;
-    uint8_t decimalPoint;
-    uint32_t stepSize;
-    uint8_t maxStringSize;
-    char stringValue[RCDEVICE_PROTOCOL_MAX_STRING_LENGTH]; // when settingType is RCDEVICE_PROTOCOL_SETTINGTYPE_INFO or RCDEVICE_PROTOCOL_SETTINGTYPE_STRING, this field store the string/info value;
-    runcamDeviceSettingTextSelection_t textSelections[RCDEVICE_PROTOCOL_MAX_TEXT_SELECTIONS];
-} runcamDeviceSettingDetail_t;
-
-typedef struct runcamDeviceWriteSettingResponse_s {
-    uint8_t resultCode;
-    uint8_t needUpdateMenuItems;
-} runcamDeviceWriteSettingResponse_t;
-
-bool runcamDeviceInit(runcamDevice_t *device);
+#define MAX_WAITING_RESPONSES 5
+
+typedef enum {
+    RCDEVICE_RESP_SUCCESS = 0,
+    RCDEVICE_RESP_INCORRECT_CRC = 1,
+    RCDEVICE_RESP_TIMEOUT = 2
+} rcdeviceResponseStatus_e;
+
+typedef struct rcdeviceResponseParseContext_s rcdeviceResponseParseContext_t;
+typedef void(*rcdeviceRespParseFunc)(rcdeviceResponseParseContext_t*);
+struct rcdeviceResponseParseContext_s {
+    uint8_t command;
+    uint8_t expectedRespLen; // total length of response data
+    uint8_t recvRespLen; // length of the data received
+    uint8_t *recvBuf; // response data buffer
+    timeMs_t timeout;
+    timeMs_t timeoutTimestamp; // if zero, it's means keep waiting for the response
+    rcdeviceRespParseFunc parserFunc;
+    runcamDevice_t *device;
+    uint8_t paramData[RCDEVICE_PROTOCOL_MAX_DATA_SIZE];
+    uint8_t paramDataLen;
+    uint8_t protocolVersion;
+    int maxRetryTimes;
+    void *userInfo;
+    rcdeviceResponseStatus_e result;
+};
+
+typedef struct {
+    uint8_t headPos; // current head position of the queue
+    uint8_t tailPos;
+    uint8_t itemCount; // the item count in the queue
+    rcdeviceResponseParseContext_t buffer[MAX_WAITING_RESPONSES];
+    rcdeviceRespParseFunc parseFunc;
+} rcdeviceWaitingResponseQueue;
+
+typedef struct rcdeviceConfig_s {
+    uint8_t initDeviceAttempts;
+    uint16_t initDeviceAttemptInterval;
+
+    // sometimes FC can't get featureInfo from devie(still no idea), so user can set it manaually.
```

Comment:
Please, apply a proper fix for the issue instead of using a setting as a workaround. Users shouldn't be manually entering parameters of the hardware when they're part of the protocol and can be automatically discovered.